File: go/exercises/exercise_01_solution.go
package main

import "fmt"

// Giải pháp: Tính tổng của một slice số nguyên
func sumSlice(numbers []int) int {
	total := 0
	for _, num := range numbers {
		total += num
	}
	return total
}

// Giải pháp: Tìm số lớn nhất trong slice
func findMax(numbers []int) int {
	if len(numbers) == 0 {
		return 0
	}
	max := numbers[0]
	for _, num := range numbers {
		if num > max {
			max = num
		}
	}
	return max
}

// Giải pháp: Đếm số chẵn trong slice
func countEven(numbers []int) int {
	count := 0
	for _, num := range numbers {
		if num%2 == 0 {
			count++
		}
	}
	return count
}

// Giải pháp: Kiểm tra số nguyên tố
func isPrime(n int) bool {
	if n < 2 {
		return false
	}
	if n == 2 {
		return true
	}
	if n%2 == 0 {
		return false
	}
	
	for i := 3; i*i <= n; i += 2 {
		if n%i == 0 {
			return false
		}
	}
	return true
}

func main() {
	// Test cases
	numbers := []int{1, 2, 3, 4, 5}
	
	if sumSlice(numbers) == 15 {
		fmt.Println("✓ sumSlice passed")
	} else {
		fmt.Println("✗ sumSlice failed")
	}
	
	if findMax(numbers) == 5 {
		fmt.Println("✓ findMax passed")
	} else {
		fmt.Println("✗ findMax failed")
	}
	
	if countEven(numbers) == 2 {
		fmt.Println("✓ countEven passed")
	} else {
		fmt.Println("✗ countEven failed")
	}
	
	if isPrime(7) && !isPrime(10) {
		fmt.Println("✓ isPrime passed")
	} else {
		fmt.Println("✗ isPrime failed")
	}
}

File: python/exercises/exercise_01_solution.py
"""
Bài tập 1: Giải pháp
"""

def sum_list(numbers):
    """Tính tổng của một danh sách số"""
    total = 0
    for num in numbers:
        total += num
    return total
    # Hoặc: return sum(numbers)

def find_max(numbers):
    """Tìm số lớn nhất trong danh sách"""
    if not numbers:
        return None
    max_num = numbers[0]
    for num in numbers:
        if num > max_num:
            max_num = num
    return max_num
    # Hoặc: return max(numbers)

def count_even(numbers):
    """Đếm số chẵn trong danh sách"""
    count = 0
    for num in numbers:
        if num % 2 == 0:
            count += 1
    return count
    # Hoặc: return len([n for n in numbers if n % 2 == 0])

def is_prime(n):
    """Kiểm tra số nguyên tố"""
    if n < 2:
        return False
    if n == 2:
        return True
    if n % 2 == 0:
        return False
    
    for i in range(3, int(n**0.5) + 1, 2):
        if n % i == 0:
            return False
    return True

# Test cases
if __name__ == "__main__":
    assert sum_list([1, 2, 3, 4, 5]) == 15
    assert find_max([1, 5, 3, 9, 2]) == 9
    assert count_even([1, 2, 3, 4, 5, 6]) == 3
    assert is_prime(7) == True
    assert is_prime(10) == False
    
    print("Tất cả test cases đã pass!")

